[ML] Add Embedding inference task type

DonalEvans opened the pull request:
This commit introduces a new `TaskType` for the inference plugin named "embedding" which allows both images and text to be used as inputs to create dense vectors. This task invokes the new `InferenceService.embeddingInfer()` method which is currently implemented to throw an `UnsupportedOperationException` if called on non-test implementations of `InferenceService`.

The input for this task is specified using a list of "content" objects, each of which specifies the type of input ("text" or "image"), the format of the input ("text" or "base64") and the String value of the input. The "format" field is optional and defaults to "text" for the "text" input type and "base64" for the "image" input type:
```
"input": [
  {
    "content": {"type": "image", "format": "base64", "value": "image data"},
  },
  {
    "content": [
      {"type": "text", "value": "text input"},
      {"type": "image", "value": "image data"}
    ]
  }
]
```
It is also possible to specify a single content object rather than a list:
```
"input": {
  "content": {"type": "text", "format": "text", "value": "text input"}
}
```
Each content object in the request will result in a single embedding array in the response, meaning that multiple texts or images can be grouped into a single content object and converted into a single embedding array if the model being used supports that.

To preserve input compatibility with the existing `text_embedding` task, the input can also be specified as a single String or a list of Strings, each of which will be internally parsed as a content object with type "text" and format "text":
```
"input": "single text input"
```
OR
```
"input": ["first text input", "second text input"]
```
The output format for this task uses the `GenericDenseEmbedding*Results` classes which produce a format like the one below (with additional "_bytes" or "_bits" suffixes on the "embeddings" array name if those element types are specified).
```
{
  embeddings=[
    {
      embedding=[54.0, ... 51.0]
    },
    {
      embedding=[45.0, ... 56.0]
    }
  ]
}
```
An `InferenceStringGroup` class is introduced to represent content objects with more than one text or image input, which is necessary to support the use case of multiple inputs generating a single embedding. This class is redundant right now, because all code paths assume a 1:1 mapping between number of inputs and number of embeddings generated, but the implementation is designed to support this use case being added in future.

This commit also adds testing for the new task type and updates existing tests and code to properly handle the new task type.

DonalEvans committed:
[ML] Add Embedding inference task type

This commit introduces a new TaskType for the inference plugin named
"embedding" which allows both images and text to be used as inputs to
create dense vectors. This task invokes the new
InferenceService.embeddingInfer() method which is currently implemented
to throw an UnsupportedOperationException if called on non-test
implementations of InferenceService.

The input for this task is specified using a list
of "content" objects, each of which specifies the type of input ("text"
or "image_base64") and the String value of the input:

"input": [
  {
    "content": {"type": "image_base64", "value": "image data"},
  },
  {
    "content": [
      {"type": "text", "value": "text input"},
      {"type": "image_base64", "value": "image data"}
    ]
  }
]

It is also possible to specify a single content object rather than a
list:

"input": {
  "content": {"type": "text", "value": "text input"}
}

Each content object in the request will result in a single embedding
array in the response, meaning that multiple texts or images can be
grouped into a single content object and converted into a single
embedding array if the model being used supports that.

To preserve input compatibility with the existing text_embedding task,
the input can also be specified as a single String or a list of Strings,
each of which will be internally parsed as a content object with type
"text":

"input": "singe text input"

OR

"input": ["first text input", "second text input"]

The output format for this task uses the GenericDenseEmbedding*Results
classes which produce a format like the one below (with additional
"_bytes" or "_bits" suffixes on the "embeddings" array name if those
element types are specified).

{
  embeddings=[
    {
      embedding=[54.0, ... 51.0]
    },
    {
      embedding=[45.0, ... 56.0]
    }
  ]
}

An InferenceStringGroup class is introduced to represent content objects
with more than one text or image input, which is necessary to support
the use case of multiple inputs generating a single embedding. This
class is redundant right now, because all code paths assume a 1:1
mapping between number of inputs and numebr of embeddings generated, but
the implementation is designed to support this use case being added in
future.

This commit also adds testing for the new task type and updates existing
tests and code to properly handle the new task type.

elasticsearchmachine commented:
Hi @donalevans, I've created a changelog YAML for you.

DonalEvans committed:
Update docs/changelog/138198.yaml

DonalEvans committed:
Fix some test failures

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Introduce NodeFeature for new task type

- In TransportInferenceUsageAction, do not return ModelStats for the
  embedding task type if some nodes in the cluster do not support the
  embedding task
- Prevent creation of inference endpoints with the embedding task type
  if some nodes in the cluster do not support the embedding task
- Add tests for the new behaviour

DonalEvans committed:
Fix line length

DonalEvans committed:
Fix unit tests

ioanatia commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapperTests.java:
```diff
@@ -439,7 +439,7 @@ private SemanticTextIndexOptions getDefaultSparseVectorIndexOptionsForMapper(Map
     public void testInvalidTaskTypes() {
         for (var taskType : TaskType.values()) {
-            if (taskType == TaskType.TEXT_EMBEDDING || taskType == TaskType.SPARSE_EMBEDDING) {
+            if (taskType == TaskType.TEXT_EMBEDDING || taskType == TaskType.SPARSE_EMBEDDING || taskType == TaskType.EMBEDDING) {
```
this test made me think initially that we made some changes to `SemanticTextFieldMapper` to support this task type. this is obviously not the case.

I wonder if instead of skipping the `TaskType.EMBEDDING` we can assert the real exception, or add at least add a comment on why we are skipping EMBEDDING here.

ioanatia commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportPutInferenceModelAction.java:
```diff
@@ -125,6 +130,18 @@ protected void masterOperation(
 
         var requestAsMap = requestToMap(request);
         var resolvedTaskType = ServiceUtils.resolveTaskType(request.getTaskType(), (String) requestAsMap.remove(TaskType.NAME));
+        if (resolvedTaskType == TaskType.EMBEDDING && featureService.clusterHasFeature(state, EMBEDDING_TASK_TYPE) == false) {
```
why do we need to do this? is it again only for the inference_usage tests when they run in bwc mode?
then I think we can just check the cluster_feature directly in the yaml test and we don't need this check here.

when we added new task types in the past (e.g. https://github.com/elastic/elasticsearch/pull/119982) we never needed a similar check

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Review comments

- Remove streaming field from EmbeddingAction.Request
- Adjust test to handle EMBEDDING task type

DonalEvans committed:
Add validator for embedding task

- Add SimpleEmbeddingServiceIntegrationValidator that's used when the
  task type is embedding
- Add tests for new class
- Add test case for embedding task to InferenceGetServicesIT

DonalEvans committed:
Merge branch 'main' into embedding-task-type

ioanatia reviewed (approved):
I'd like to get more eyes on this from @jonathan-buttner or maybe @timgrein that have more context on the inference API.

One thing I want to note is that at this stage it is not necessary IMO to support both formats:

```
"input": [
  {
    "content": {"type": "image_base64", "value": "image data"},
  },
  {
    "content": [
      {"type": "text", "value": "text input"},
      {"type": "image_base64", "value": "image data"}
    ]
  }
]
```
and
```
"input": {
  "content": {"type": "text", "value": "text input"}
}
```
I would just go with the first one for now since it is the more generic one.
It is a bit premature to support both at this stage.
But I don't want to block progress here - we can discuss this separately.

jonathan-buttner commented on server/src/test/java/org/elasticsearch/inference/EmbeddingRequestTests.java:
```diff
@@ -0,0 +1,179 @@
+
+import static org.hamcrest.Matchers.is;
+
+public class EmbeddingRequestTests extends AbstractWireSerializingTestCase<EmbeddingRequest> {
```
Can we use the bwc ones here? That way if/when we add transport version checks we don't forget to switch it then.

jonathan-buttner commented on server/src/test/java/org/elasticsearch/inference/InferenceStringTests.java:
```diff
@@ -0,0 +1,157 @@
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+
+public class InferenceStringTests extends AbstractWireSerializingTestCase<InferenceString> {
```
How about we make this one bwc as well?

jonathan-buttner reviewed (commented):
Great work Donal. I left a few questions and suggests around tests and validation. Looks great though.

DonalEvans committed:
Review comments

- Make InferenceString and InferenceStringGroup Writeable instead of
  NamedWriteable
- Add missing test classes
- Convert existing tests to backwards compatibility tests
- Provide more descriptive exception message on parser error

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Rework API design to separate input type from format

- Rename IMAGE_BASE64 to IMAGE
- Introduce DataFormat enum in InferenceString
- Add and update tests

DonalEvans committed:
Add missing comma

DonalEvans committed:
Merge branch 'main' into embedding-task-type

DonalEvans committed:
Merge branch 'main' into embedding-task-type
